modules/dns-helper/dns_local_server.go

```diff
@@ -0,0 +1,160 @@
+	Server           *dns.Server
+	DNSDatabase      dnsDatabase
+	DNSDatabaseRetry dnsDatabase
+}
+
+// newDNSTestServer returns a new instance of dnsTestServer
+func newDNSTestServer(server *dns.Server) *dnsTestServer {
+	return &dnsTestServer{Server: server, DNSDatabase: make(dnsDatabase), DNSDatabaseRetry: make(dnsDatabase)}
+}
+
+// Address returns the host:port string of the server listener
+func (s *dnsTestServer) Address() string {
+	return s.Server.PacketConn.LocalAddr().String()
+}
+
+// AddEntryToDNSDatabase adds DNSAnswers to the DNSQuery in the database of the server
+func (s *dnsTestServer) AddEntryToDNSDatabase(q DNSQuery, a DNSAnswers) {
+	s.DNSDatabase[q] = append(s.DNSDatabase[q], a...)
+}
+
+// AddEntryToDNSDatabaseRetry adds DNSAnswers to the DNSQuery in the database used when retrying
+func (s *dnsTestServer) AddEntryToDNSDatabaseRetry(q DNSQuery, a DNSAnswers) {
+	s.DNSDatabaseRetry[q] = append(s.DNSDatabaseRetry[q], a...)
+}
+
+// setupTestDNSServers runs and returns 2x local dnsTestServer, initialized with NS records for the testDomain pointing to themselves
+// it uses a handler that will send replies stored in their internal DNSDatabase
+func setupTestDNSServers() (s1, s2 *dnsTestServer) {
+	s1 = runTestDNSServer("0")
+	s2 = runTestDNSServer("0")
+
+	q := DNSQuery{"NS", testDomain}
+	a := DNSAnswers{{"NS", s1.Address() + "."}, {"NS", s2.Address() + "."}}
+	s1.AddEntryToDNSDatabase(q, a)
+	s2.AddEntryToDNSDatabase(q, a)
+
+	s1.Server.Handler.(*dns.ServeMux).HandleFunc(testDomain+".", func(w dns.ResponseWriter, r *dns.Msg) {
+		stdDNSHandler(w, r, s1, false)
+	})
+	s2.Server.Handler.(*dns.ServeMux).HandleFunc(testDomain+".", func(w dns.ResponseWriter, r *dns.Msg) {
+		stdDNSHandler(w, r, s2, true)
+	})
+
+	return s1, s2
+}
+
+// setupTestDNSServersRetry runs and returns 2x local dnsTestServer, initialized with NS records for the testDomain pointing to themselves
+// it uses a handler that will send replies stored in their internal DNSDatabase, and then switch to their DNSDatabaseRetry after some time
+func setupTestDNSServersRetry() (s1, s2 *dnsTestServer) {
+	s1 = runTestDNSServer("0")
+	s2 = runTestDNSServer("0")
+
+	q := DNSQuery{"NS", testDomain}
+	a := DNSAnswers{{"NS", s1.Address() + "."}, {"NS", s2.Address() + "."}}
+	s1.AddEntryToDNSDatabase(q, a)
+	s2.AddEntryToDNSDatabase(q, a)
+	s1.AddEntryToDNSDatabaseRetry(q, a)
+	s2.AddEntryToDNSDatabaseRetry(q, a)
+
+	s1.Server.Handler.(*dns.ServeMux).HandleFunc(testDomain+".", func(w dns.ResponseWriter, r *dns.Msg) {
+		retryDNSHandler(w, r, s1, false)
+	})
+	s2.Server.Handler.(*dns.ServeMux).HandleFunc(testDomain+".", func(w dns.ResponseWriter, r *dns.Msg) {
+		retryDNSHandler(w, r, s2, true)
+	})
+
+	return s1, s2
+}
+
+// runTestDNSServer starts and returns a new dnsTestServer listening in localhost and the given UDP port
+func runTestDNSServer(port string) *dnsTestServer {
+	listener, err := net.ListenPacket("udp", "127.0.0.1:"+port)
+
+	if err != nil {
+		log.Fatal(err)
```

Comment:
As mentioned in the other comment, we want to avoid `log.Fatal` in tests. You should switch this to `t.Fatal`, which will basically do the same thing but terminate the test goroutine instead of the whole process.

Reply:
Switch from `log.Fatal` to `t.Fatal`